### DISCUSSION TITLE
Gatsby Images are only being displayed/found once every few builds

### DISCUSSION THREAD

**Discussion by JackDevAU** [gatsbyjs/gatsby#34408] [Help]
Hi All,

#### Package versions:
**gatsby**: "^3.14.6",
**gatsby-plugin-remote-images:** "^3.2.0",
**gatsby-source-custom-api:** "2.2.1",
**gatsby-plugin-image:** "^1.14.2",
**gatsby-transformer-sharp:** "^3.14.0",
**gatsby-plugin-sharp:** "^3.14.3",

### My Issue
I'm having an issue where images that are pulled in from my API are not always being found when querying. Currently I've tried a few different approaches to try and fix this issue but it seems to be that once the images are cached in the `.cache` folder the images (all images) will only be found once every few builds...

My first approach was to try and use the Gatsby plugin `gatsby-source-custom-api`
```JavaScript
	{
		resolve: "gatsby-source-custom-api",
		options: {
			url: `${process.env.GATSBY_API_BASE_URL}/${apiRoute}`,
			rootKey: graphEntityName,
			schemas: graphSchema,
			imageKeys: ["thumbnails"],
		},
	},
```
**Figure: under plugins -  gatsby-config.js file**
<br />
This didn't work due to the image **url** being another call to the api `http://localhost:5000/api/news/<newsid here>/thumbnail` which will download the image. This threw a warning `Image-Extension not valid`

The next attempt was to use the `gatsby-plugin-remote-images` plugin. This successfully downloads the images and creates a field/node under the parent node called `thumbnailImages`. 

```JavaScript
	{
		resolve: `gatsby-plugin-remote-images`,
		options: {
			nodeType: graphEntityName,
			imagePath: imagePath,
			name: "thumbnailImages",
		},
	},
```
**Figure: under plugins -  gatsby-config.js file**

<br  />

The query using this method is as follows:

```JavaScript
query MyQuery {
  allEventItem {
    nodes {
      thumbnailImages {
        childImageSharp {
          fluid {
            src
          }
        }
      }
    }
  }
}
```

For some reason this also creates multiple of the same fields on the query... though I believe this to be an issue with the gatsby-source-custom-api` plugin.  When it does display images only the middle field is populated.

```JavaScript
{
  "data": {
    "allEventItem": {
      "nodes": [
        {
          "thumbnailImages": []
        },
        {
          "thumbnailImages": []
        },
        {
          "thumbnailImages": []
        }
      ]
    }
  },
  "extensions": {}
}
```
**Figure: Output from Graphql explorer**
<br />
Due to the randomness with this I tried another method which is using the `onNodeCreate` and uses the `createRemoteFileNode` to add a field to the current parent node. This will be either a "newsItem" or an "eventItem"

```JavaScript
const { createRemoteFileNode } = require(`gatsby-source-filesystem`)

exports.onCreateNode = async ({
	node,
	actions,
	store,
	cache,
	reporter,
	createNodeId,
}) => {
	const { createNode, createNodeField } = actions

	if (node.internal.type === "newsItem" || node.internal.type === "eventItem") {
		try {
			const fileNode = await createRemoteFileNode({
				url: node.thumbnails[0]?.url,
				parentNodeId: node.id,
				store,
				cache,
				reporter,
				createNode,
				createNodeId,
			})

			if (fileNode) {
				createNodeField({ node, name: "thumbnailImages", value: fileNode.id })
			}
		} catch (e) {}
	}
}

exports.createSchemaCustomization = ({ actions }) => {
	const { createTypes } = actions

	createTypes(`
		type eventItem implements Node {
		thumbnailImages: File @link(from: "fields.thumbnailImages")
		}
	`)

	createTypes(`
		type newsItem implements Node {
			thumbnailImages: File @link(from: "fields.thumbnailImages")
		}
	`)
}
```
**Figure: gatsby-node.js file**

<br  />

However, as with the others this only gets the image once every few builds AFTER the image has been cached. The first build will always show the images however after that its once every few.... 

```JavaScript
{
  "data": {
    "allEventItem": {
      "nodes": [
        {
          "thumbnailImages": null
        },
        {
          "thumbnailImages": {
            "childImageSharp": {
              "fluid": {
                "src": "/static/934fb6a570c47c42ab3bedb0b225b20e/ee604/thumbnail.png"
              }
            }
          }
        },
        {
          "thumbnailImages": null
        }
      ]
    }
  },
  "extensions": {}
}
```
**Figure: Output from Graphql explorer**
<br />
using `gatsby clean` before building does solve the issue... however, on Gatsby cloud I don't have the option with automatic builds. 

Any help at all would be greatly appreciated!

**Comment by joshdavenport** [comment]
Did you ever get to the bottom of this one @JackDevAU ?

**Reply by JackDevAU** [reply]
Hey @joshdavenport,

Unfortunately, no. This got put on the backburner. 

The next steps would be to try recreate the issue in a new project and see if newer or older versions of the plugins/Gatsby have the same issues. However, I'm not sure when I'll have time to continue to look into this.
